src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+﻿﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+namespace Microsoft.Quantum.QsCompiler
+{
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+    // Type Parameters are frequently referenced by the callable of the type parameter followed by the name of the specific type parameter.
+    using TypeParameterName = Tuple<QsQualifiedName, NonNullable<string>>;
+    using TypeParameterResolutions = ImmutableDictionary</*TypeParameterName*/ Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+
+    /// <summary>
+    /// Summarizes a series of type parameter resolution dictionaries, ResolutionsInSummary, into one
+    /// resolution dictionary, CombinedTypeParameterResolutions, containing the ultimate type resolutions
+    /// for all the type parameters found in the dictionaries. Validation is done on the resolutions,
+    /// which can be checked through the IsValidSummary, SummarizesOverConflictingResolution, and
+    /// SummarizesOverParameterConstriction flags.
+    /// </summary>
+    public class TypeResolutionSummary
```

Comment:
I am not totally convinced about the use of the word summary since I think it is more commonly associated with brief. Maybe something like `TypeResolutionAggregate` or `TypeResolutionBundle` or `TypeResolutionCollection`. Maybe it could be more explicit like `CombinedTypeResolution` (which is my preferred one so far). Naming is hard but I think there might be better options than summary in this case.

Reply:
That's disappointing to hear. I thought the summary narrative made this whole concept come together nicely and made things easier to understand. It certainly made all of this more cohesive for me. I am concerned that this class will be more confusing without understanding the `CombinedTypeParameterResolutions` field is just a summary over the `ResolutionsInSummary`. Are you sure I should get rid of this way of conceptualizing the class?

Reply:
I'm also not really familiar with this use of the word summary, and "summarizing over" something sounds strange to me - I had to read the doc comments to understand it. I would describe something as "summarizing" something or being a "summary of" something but I'm not sure that applies here.

I agree with @cesarzc that it seems like you could describe this as merging or combining separate type parameter mappings into one mapping.

Reply:
The term 'summary' has been replaced with the term 'combination'.